### ISSUE TITLE
Magento 2.2.1 - Impossible to create a new storeview

### ISSUE THREAD

**Issue by hostep** [magento/magento2#12405]
### Preconditions
1. Magento 2.2.1
2. PHP 7.0.25

### Steps to reproduce
1. Setup a clean Magento 2.2.1
2. Install https://github.com/baldwin-agency/magento2-module-less-js-compiler
3. After running `setup:upgrade` check your `app/etc/config.php` file, it will have `Magento_Config` as a higher priority to load then `Magento_Store`
4. Login into the backend
5. Go to Stores => All Stores
6. Click 'Create Store View'
7. Enter some dummy content in the fields and click 'Save Store View'
8. An error `Requested store is not found` is shown and no storeview is added.

### Expected result
1. A new storeview should be created

### Actual result
1. An error `Requested store is not found` is shown and no storeview is added.

**Comment by hostep** [comment]
Creating a Website is also not working, error: `No such entity.`
Creating a Store is actually working, surprisingly.

**Comment by hostep** [comment]
Sorry, I lied in my initial post, it wasn't on a clean installation. Using a clean installation it actually works. Now to figure out why it didn't work in my case (an "almost" clean installation).

Closing for now, sorry!

**Comment by hostep** [comment]
Ok, in case anyone is interested, this was because we had a simple module installed which depended on `Magento_Developer`.
When running `setup:upgrade` the order of modules in the `app/etc/config.php` file were shuffled and for some reason `Magento_Config` was inserted higher in the list then `Magento_Store`. This wrong ordering then caused the problem.

I still think this is an actual bug in Magento, because the `Magento_Config` module depends on functionality of `Magento_Store` but there isn't any explicit dependency set in `Magento_Config`'s [`module.xml` file](https://github.com/magento/magento2/blob/2.2.1/app/code/Magento/Config/etc/module.xml).
When I add this manually in that file:
```xml
        <sequence>
            <module name="Magento_Store"/>
        </sequence>
```

Then run `setup:upgrade` again, it works!

I'll create a PR for this shortly.

**Comment by hostep** [comment]
Reopening.

I updated the steps to reproduce in my initial post with exact steps about how you can reproduce the problem.
I'm not convinced anymore that adding `Magento_Store` in the `sequence` of the `Magento_Config` module is the correct fix.

While searching the root cause of this problem, I went through a million classes which have something to do with Caching and especially with cleaning the cache, but I'm stuck debugging this, got to the class `Cm_Cache_Backend_File` and I'm no longer convinced I'm searching in the right direction. Maybe time to find a proper IDE instead of putting `die` and `var_dump` statements all over the place. If someone with better debugging skills can find the real cause that would be appreciated!

----
Bonus

While going through a bunch of classes, I noticed [a call to the method `clean`](https://github.com/magento/magento2/blob/2.2.1/app/code/Magento/Store/Model/StoreManager.php#L238) on an [instance of a class](https://github.com/magento/magento2/blob/2.2.1/lib/internal/Magento/Framework/App/Config.php) which implements [ScopeConfigInterface](https://github.com/magento/magento2/blob/2.2.1/lib/internal/Magento/Framework/App/Config/ScopeConfigInterface.php) while that interface has no definition of the `clean` method. This feels wrong to me, no?

**Comment by magento-engcom-team** [comment]
@hostep, thank you for your report.
We've created internal ticket(s) MAGETWO-84636 to track progress on the issue.

**Comment by leek** [comment]
@hostep You don't happen to be using the [AvS_ScopeHint2](https://github.com/avstudnitz/AvS_ScopeHint2) extension on your store that is experiencing this issue, are you? We we're experiencing this exact same problem and after removing that module, it was resolved.

**Comment by hostep** [comment]
@leek: no, it was on a clean installation with one extra module installed as mentioned in my initial post. This extra module shuffles the sort order around in which other modules are loaded, which causes the issue. I'm assuming the same thing happens to you.
It might as well be fixed by installing other modules. Since the sort order of modules in the `config.php` file isn't really deterministic for modules with the same weight as others, it's more like a random shuffle based on no real logic.

**Comment by hamdrew** [comment]
I encountered this issue of a seemingly randomly-shuffled **config.php** file. It turns out that my module's **module.xml** file was invalid.  After fixing it, the **config.php** was ordered correctly.

Ensure that the `<sequence>` tags are in the correct location for all your modules:

```xml
<!-- INVALID -->
<module name="Hamdrew_Module" setup_version="1.0.0"></module>
<sequence>
    <module name="Magento_Catalog" />
</sequence>

<!-- VALID -->
<module name="Hamdrew_Module" setup_version="1.0.0">
    <sequence>
        <module name="Magento_Catalog" />
    </sequence>
</module>
```

I didn't catch this until I took the time to configure the [URN Highlighter](http://devdocs.magento.com/guides/v2.0/config-guide/cli/config-cli-subcommands-urn.html) in IntelliJ/PHPStorm.

**Comment by magento-engcom-team** [comment]
Hi @hostep. Thank you for your report. 
The issue has been fixed in magento/magento2#13943 by @hostep in 2.2-develop branch
Related commit(s):
 - [5b6df42840edf7d27455fa6fe8dec221cf8af747](https://github.com/magento/magento2/commit/5b6df42840edf7d27455fa6fe8dec221cf8af747)

The fix will be available with the upcoming 2.2.5 release.

**Comment by magento-engcom-team** [comment]
Hi @hostep. Thank you for your report. 
The issue has been fixed in magento/magento2#14043 by @hostep in 2.3-develop branch
Related commit(s):
 - [7a2d71561fbde5ea849fbeccaf2847fcddd6adc6](https://github.com/magento/magento2/commit/7a2d71561fbde5ea849fbeccaf2847fcddd6adc6)

The fix will be available with the upcoming 2.3.0 release.
